notifier/processor.go

```diff
@@ -0,0 +1,231 @@
+		return nil
+	}
+
+	notifications := []Notification{}
+	create := func(r Reason, affected claircore.AffectedManifests) error {
+		for manifest, vulns := range affected.VulnerableManifests {
+			// summarize most severe vuln affecting manifest
+			// the vulns array will be sorted by most severe
+			var vs VulnSummary
+			vuln := affected.Vulnerabilities[vulns[0]]
+			vs.FromVulnerability(*vuln)
+
+			digest, err := claircore.ParseDigest(manifest)
+			if err != nil {
+				return err
+			}
+			n := Notification{
+				Manifest:      digest,
+				Reason:        r,
+				Vulnerability: vs,
+			}
+			notifications = append(notifications, n)
+		}
+		return nil
+	}
+	err = create(Added, added)
+	if err != nil {
+		return err
+	}
+	err = create(Removed, removed)
+	if err != nil {
+		return err
+	}
+	opts := PutOpts{
+		Updater:        e.updater,
+		UpdateID:       e.uo.Ref,
+		NotificationID: uuid.New(),
+		Notifications:  notifications,
+	}
+	err = p.store.PutNotifications(ctx, opts)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// safe guards against situations where creating notifications is
+// incorrect.
+//
+// will be performed under a distributed lock.
+func (p *Processor) safe(ctx context.Context, e Event) bool {
+	uoid := e.uo.Ref.String()
+	log := zerolog.Ctx(ctx).With().
+		Uint8("processor_id", p.id).
+		Str("component", "notifier/processor/Processor.Process").
+		Str("updater", e.updater).
+		Str("UOID", uoid).
+		Logger()
+	// confirm we are not making duplicate notifications
+	var errNoReceipt clairerror.ErrNoReceipt
+	_, err := p.store.ReceiptByUOID(ctx, e.uo.Ref)
+	switch {
+	case errors.As(err, &errNoReceipt):
+		// hop out of switch
+	case err != nil:
+		log.Error().Err(err).Msg("received error getting receipt by UOID")
+		return false
+	default:
+		log.Info().Msg("receipt created by another processor. will not process notifications")
+		return false
+	}
+
+	// confirm UOID is not stale and get previous UOID for diffing if exists.
+	// if no previous UOID, return false, we don't want a full diff of notifications
+	// TODO(louis) UpdateOperations signature supports getting "all" for a given updater
+	// but code path is not implemented. implement this to optimize.
+	all, err := p.matcher.UpdateOperations(ctx)
+	if err != nil {
+		log.Error().Err(err).Msg("received error getting update operations from matcher")
+		return false
+	}
+	if _, ok := all[e.updater]; !ok {
+		log.Warn().Msg("updater missing from update operations returned from matcher. matcher may have garbage collected")
+		return false
+	}
+
+	uos := all[e.updater]
+	n := len(uos)
+	if n < 2 {
+		log.Info().Msg("encountered first update operation. will not process notifications")
+		return false
+	}
+
+	current, prev := uos[0], uos[1]
+	if current.Ref.String() != e.uo.Ref.String() {
+		log.Info().Str("new", current.Ref.String()).Msg("newer update operation is present, will not process notifications")
+		return false
+	}
+	p.previous = prev.Ref
```

Comment:
the safe method is designed to just return a bool on whether or not the processor should continue processing.

Reply:
OK, but it actually returns `(ok, previousRef)`, it just does the second one via assigning it to shared state instead of returning it.

Having this method return it and the `create` method take it would make the sequence dependency explicit.

Reply:
I changed it.